File: PyTorch_GPU/main_algorithm.py
import torch
import time

def dgemm_pytorch(A, B):
    """
    Performs matrix multiplication using PyTorch on the GPU and returns the execution time.
    """
    # Ensure tensors are on the same device, although we create them on the device directly.
    # This is good practice if the function were to be used elsewhere.
    device = A.device
    
    # Synchronize before timing to ensure all previous CUDA operations are complete.
    torch.cuda.synchronize(device=device)
    
    start_time = time.time()
    
    # Perform the matrix multiplication
    C = torch.matmul(A, B)
    
    # Synchronize after timing to wait for the matmul operation to finish.
    torch.cuda.synchronize(device=device)
    
    end_time = time.time()
    
    return C, end_time - start_time

if __name__ == '__main__':
    # Check if a CUDA-enabled GPU is available
    if not torch.cuda.is_available():
        print("CUDA is not available. This script requires a GPU.")
        exit()

    device = torch.device("cuda")
    print(f"Using device: {torch.cuda.get_device_name(device)}")

    time_spent = 0.0
    end_time = 1 * 60  # 1 minute in seconds
    n = 512
    multiplication_count = 0

    print(f"Performing matrix multiplications of size {n}x{n} for {end_time} seconds on {device}.")

    try:
        # Generate matrices once before the loop to avoid repeated allocation overhead
        A = torch.randn(n, n, device=device, dtype=torch.float64)
        B = torch.randn(n, n, device=device, dtype=torch.float64)

        while time_spent < end_time:
            # Perform matrix multiplication and get the time
            C, mult_time = dgemm_pytorch(A, B)
            
            time_spent += mult_time
            multiplication_count += 1

        # Clean up tensors to free GPU memory after the loop
        del A, B, C
        torch.cuda.empty_cache()

    except torch.cuda.OutOfMemoryError:
        print(f"GPU out of memory for n = {n}. Halting.")
        # Clean up memory before breaking
        torch.cuda.empty_cache()
    except Exception as e:
        print(f"An error occurred: {e}")

    print(f"\n-------------------------------------------------")
    print(f"Fixed N: {n}")
    print(f"Number of multiplications performed: {multiplication_count}")
    print(f"Total computation time: {time_spent:.2f} seconds")
    print(f"-------------------------------------------------")

File: MKL/main_algorithm.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "mkl.h"

int main() {
    double time_spent = 0.0;
    double end_time = 1 * 60; // 1 minute in seconds
    int n = 512;
    int multiplication_count = 0;

    printf("Starting MKL DGEMM benchmark with fixed N = %d for %.f seconds.\n", n, end_time);

    double *A, *B, *C;
    double alpha = 1.0, beta = 0.0;

    // MKL recommends using mkl_malloc for better memory alignment
    A = (double *)mkl_malloc(n * n * sizeof(double), 64);
    B = (double *)mkl_malloc(n * n * sizeof(double), 64);
    C = (double *)mkl_malloc(n * n * sizeof(double), 64);

    if (A == NULL || B == NULL || C == NULL) {
        fprintf(stderr, "Memory allocation failed for n = %d\n", n);
        if (A) mkl_free(A);
        if (B) mkl_free(B);
        if (C) mkl_free(C);
        return 1;
    }

    // Initialize matrices A and B with random values
    for (int i = 0; i < (n * n); i++) {
        A[i] = (double)(i + 1);
        B[i] = (double)(n * n - i);
    }

    while (time_spent < end_time) {
        double start_dgemm, end_dgemm;
        
        // --- Time the dgemm operation ---
        start_dgemm = dsecnd();

        cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans,
                    n, n, n, alpha, A, n, B, n, beta, C, n);

        end_dgemm = dsecnd();
        // --------------------------------

        time_spent += end_dgemm - start_dgemm;
        multiplication_count++;
    }

    // Free memory
    mkl_free(A);
    mkl_free(B);
    mkl_free(C);

    printf("\n-------------------------------------------------\n");
    printf("Fixed N: %d\n", n);
    printf("Number of multiplications performed: %d\n", multiplication_count);
    printf("Total computation time: %.2f seconds\n", time_spent);
    printf("-------------------------------------------------\n");

    return 0;
}